### PR TITLE
9241 new fixed and diverging palettes

### PR DESCRIPTION
**Pull request by russellburdt**
This pull request is related to Issue [9241](https://github.com/bokeh/bokeh/issues/9241).

Fixed 256-length palettes were added and a diverging palette generator function was added.  A single test was added to test_palettes.py.

### CODE REVIEW

**Comment by bryevdv** [comment]
@russellburdt The integration test failure is spurios, but there is one real failure on Py2.7, not sure the immediate cause:

```
_________________________ test_cmap_generator_function _________________________
    def test_cmap_generator_function():
        assert pal.viridis(256) == pal.Viridis256
        assert pal.magma(256) == pal.Magma256
        assert pal.plasma(256) == pal.Plasma256
        assert pal.inferno(256) == pal.Inferno256
        assert pal.gray(256) == pal.Greys256
        assert pal.grey(256) == pal.Greys256
        assert pal.turbo(256) == pal.Turbo256
>       assert pal.diverging_palette(pal.Reds9, pal.Greys9, n=18, midpoint=0.5) == pal.Reds9 + pal.Greys9[::-1]
bokeh/tests/test_palettes.py:44: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
bokeh/palettes.py:1370: in diverging_palette
    return self.linear_palette(palette1, n1) + self.linear_palette(palette2, n2)
bokeh/palettes.py:1336: in linear_palette
    return [ palette[int(math.floor(i))] for i in np.linspace(0, len(palette)-1, num=n) ]
../../../miniconda/envs/testenv/lib/python2.7/site-packages/numpy/core/function_base.py:128: in linspace
    num = _index_deprecate(num)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
i = 9, stacklevel = 3
    def _index_deprecate(i, stacklevel=2):
        try:
            i = operator.index(i)
        except TypeError:
            msg = ("object of type {} cannot be safely interpreted as "
                   "an integer.".format(type(i)))
            i = int(i)
            stacklevel += 1
>           warnings.warn(msg, DeprecationWarning, stacklevel=stacklevel)
E           TypeError: expected string or buffer
../../../miniconda/envs/testenv/lib/python2.7/site-packages/numpy/core/function_base.py:28: TypeError
```

Also some linting issues were flagged:
```
E         ./bokeh/palettes.py:1177:166: E501 line too long (188 > 165 characters)
E         ./bokeh/palettes.py:1178:166: E501 line too long (186 > 165 characters)
E         ./bokeh/palettes.py:1179:166: E501 line too long (187 > 165 characters)
E         ./bokeh/palettes.py:1180:166: E501 line too long (188 > 165 characters)
E         ./bokeh/palettes.py:1181:166: E501 line too long (185 > 165 characters)
```

File: bokeh/palettes.py [hunk 1]
```diff
@@ -1054,11 +1174,11 @@ def PuRd(self):     return { 3: self.PuRd3,     4: self.PuRd4,     5: self.PuRd5
...
-    def Reds(self):     return { 3: self.Reds3,     4: self.Reds4,     5: self.Reds5,     6: self.Reds6,     7: self.Reds7,     8: self.Reds8,     9: self.Reds9 }
+    def Purples(self):  return { 3: self.Purples3,  4: self.Purples4,  5: self.Purples5,  6: self.Purples6,  7: self.Purples7,  8: self.Purples8,  9: self.Purples9,  256: self.Purples256 }
+    def Blues(self):    return { 3: self.Blues3,    4: self.Blues4,    5: self.Blues5,    6: self.Blues6,    7: self.Blues7,    8: self.Blues8,    9: self.Blues9,    256: self.Blues256 }
+    def Greens(self):   return { 3: self.Greens3,   4: self.Greens4,   5: self.Greens5,   6: self.Greens6,   7: self.Greens7,   8: self.Greens8,   9: self.Greens9,   256: self.Greens256 }
+    def Oranges(self):  return { 3: self.Oranges3,  4: self.Oranges4,  5: self.Oranges5,  6: self.Oranges6,  7: self.Oranges7,  8: self.Oranges8,  9: self.Oranges9,  256: self.Oranges256 }
+    def Reds(self):     return { 3: self.Reds3,     4: self.Reds4,     5: self.Reds5,     6: self.Reds6,     7: self.Reds7,     8: self.Reds8,     9: self.Reds9,     256: self.Reds256 }
```
Comment (bryevdv) [inline bokeh/palettes.py]: suggest just adding the `# NOQA` comment like other long lines in this module

**Comment by bryevdv** [comment]
@russellburdt FYI in the interest of merging this ASAP I made the small changes here in the GH UI directly

**Comment by russellburdt** [comment]
@bryevdv Thank you for adding the int casts and getting this merged.  With your changes are all tests passing now?

**Comment by bryevdv** [comment]
@russellburdt everything just went green but I just noticed that the module docstring was not updated, which means that the new function will not appear in the docs. Since this module has some special requirements, everything is called out explicitly. So a new entry needs to be added here:
```
Functions
---------

The ``bokeh.palettes`` module also has several functions that can be used
to generate palettes of arbitrary size.

.. autofunction:: bokeh.palettes.cividis(n)
.. autofunction:: bokeh.palettes.gray(n)
.. autofunction:: bokeh.palettes.grey(n)
.. autofunction:: bokeh.palettes.inferno(n)
.. autofunction:: bokeh.palettes.linear_palette(palette, n)
.. autofunction:: bokeh.palettes.magma(n)
.. autofunction:: bokeh.palettes.viridis(n)
```

Additionally, `n` is not optional in any of the other functions in this module. For consistency I thing the default value of 256 should be removed. Also, the `midpoint` docstring entry needs to be updated to state "option" as well as provide the default value.

File: bokeh/palettes.py [hunk 2]
```diff
@@ -1215,6 +1335,40 @@ def linear_palette(self, palette, n):
             raise ValueError("Requested %(r)s colors, function can only return colors up to the base palette's length (%(l)s)" % dict(r=n, l=len(palette)))
         return [ palette[int(math.floor(i))] for i in np.linspace(0, len(palette)-1, num=n) ]
 
+    def diverging_palette(self, palette1, palette2, n=256, midpoint=0.5):
```
Comment (bryevdv) [inline bokeh/palettes.py]: `n` is not optional in any of the other functions, it should be made to match

File: bokeh/palettes.py [hunk 3]
```diff
@@ -1215,6 +1335,40 @@ def linear_palette(self, palette, n):
...
+        left to right from perceptually dark to light colors.  In that case the returned
+        palette is comprised of the input palettes connected at perceptually light ends.
+        Palettes are combined by piecewise linear interpolation.
+
+        Args:
+
+            palette1 (list[str]) : a list of hex RGB color strings
+            palette2 (list[str]) : a list of hex RGB color strings
+            n (int) : the size of the output palette to generate
+            midpoint (float) : relative position in the returned palette where input palettes are connected
```
Comment (bryevdv) [inline bokeh/palettes.py]: needs `(float, optional)` for type and `(default: 0.5)` at the end

**Comment by bryevdv** [comment]
Thanks @russellburdt

**Pull request merged**